styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/AbstractRunnerHandler.java

```diff
@@ -0,0 +1,79 @@
+/*-
+ * -\-\-
+ * Spotify Styx Scheduler Service
+ * --
+ * Copyright (C) 2016 - 2020 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.styx.state.handlers;
+
+import com.spotify.styx.model.Event;
+import com.spotify.styx.model.ExecutionDescription;
+import com.spotify.styx.state.EventRouter;
+import com.spotify.styx.state.OutputHandler;
+import com.spotify.styx.state.RunState;
+import com.spotify.styx.state.StateData;
+import java.util.function.Predicate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An abstract {@link OutputHandler}  starts docker runs on
+ * {@link RunState.State#SUBMITTED} transitions
+ */
+abstract class AbstractRunnerHandler implements OutputHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractRunnerHandler.class);
+  private final Predicate<ExecutionDescription> applicableFn;
+
+  protected AbstractRunnerHandler(Predicate<ExecutionDescription> applicableFn) {
+    this.applicableFn = applicableFn;
+  }
+
+  @Override
+  public final void transitionInto(RunState state, EventRouter eventRouter) {
+    var stateData = state.data();
+    switch (state.state()) {
+      case SUBMITTING:
+      case SUBMITTED:
+      case RUNNING:
+        var workflowInstance = state.workflowInstance();
+        if (stateData.executionId().isEmpty()) {
+          LOG.error("Unable to start procedure. Missing execution id for " + workflowInstance);
+          eventRouter.receiveIgnoreClosed(Event.halt(workflowInstance), state.counter());
+          return;
+        }
+        var maybeExecutionDescription = stateData.executionDescription();
+        if (maybeExecutionDescription.isEmpty()) {
```

Comment:
*To capture a comment from our hangout conversation*
Is there any way around introducing this new requirement of having an ExecutionDescription in the SUBMITTING and RUNNING state to distinguish when to execute the Flyte vs the Docker-RunnerHandler

Reply:
RunState is not mutated by any handler. It is just mutated form computing the [nextRunState](https://github.com/spotify/styx/blob/b68cdb17612182bf6f8d869b539ca53b62efae3d/styx-scheduler-service/src/main/java/com/spotify/styx/state/PersistentStateManager.java#L213). This use the visitor pattern and at the end calls [the methods](https://github.com/spotify/styx/blob/master/styx-service-common/src/main/java/com/spotify/styx/state/RunState.java#L124) on the TransitionVisitor from RunState. There we can see that we only append new info to the RunState, so any information present at one stae in the state machine would be present on downstream states. In this particular case, executionDescription and executionId checks on SUBMITTING will continue to be valid in SUBMITTED and RUNNING, **so it is safe to do this**

Reply:
Adding tests for this